optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/tri/TriConstraintStream.java

```diff
@@ -196,6 +198,80 @@
         return join(otherClass, AbstractQuadJoiner.merge(joiners));
     }
 
+    // ************************************************************************
+    // Group by
+    // ************************************************************************
+
+    /**
+     * Runs all tuples of the stream through a given @{@link TriConstraintCollector} and converts them into a new
+     * {@link UniConstraintStream} which only has a single tuple, the result of applying {@link TriConstraintCollector}.
+     * @param collector never null, the collector to perform the grouping operation with
+     * @param <ResultContainer_> the mutable accumulation type (often hidden as an implementation detail)
+     * @param <Result_> the type of a fact in the destination {@link UniConstraintStream}'s tuple
+     * @return never null
+     */
+    <ResultContainer_, Result_> UniConstraintStream<Result_> groupBy(
+            TriConstraintCollector<A, B, C, ResultContainer_, Result_> collector);
+
+    /**
+     * Convert the {@link TriConstraintStream} to a {@link UniConstraintStream}, containing the set of tuples resulting
+     * from applying the group key mapping function on all tuples of the original stream. Neither tuple of the new
+     * stream will {@link Objects#equals(Object, Object)} any other.
```

Comment:
avoid future tense "will".
Nice to have: in a javadoc, if a line ends, start a new line for the next one (so \n before between "stream." and "Neither". The html puts them together anyway, but this gives us "ventilated proze", which is easier to author and maintain.